Feature/pok 29 hide pokémon spawns

Bischi1997 opened the pull request.

Bischi1997 committed:
added hideEventSpawnfunctionality

Bischi1997 committed:
updated error on function convertDate()

added +1 for month, as the month starts at zero, but january should be 01

Bischi1997 committed:
updated pokemon-detail.components.ts for dev to pok-29 merge

Bischi1997 committed:
Merge remote-tracking branch 'origin/development' into feature/POK-29-hide-pokémon-spawns

Bischi1997 committed:
added check for onlyeventexclusive spawns

Finrod-Amandil commented:
For Pokémon which do not have any inactive, event-exclusive spawns (i.e. bulbasaur), the button "Show/Hide inactive event-exclusive spawns" should not be show, as it has no effect.

Finrod-Amandil commented:
When showing inactive event-exclusive spawns, apply the sorting again. Currently, the list is no longer correctly sorted after adding the event-exclusive spawns again.
Steps to reproduce: 
- Go to page of corphish (1 normal spawn, 1 event exclusive)
- Show event-exclusive spawns
- Sort list by location descending, so that the event-exclusive spawn is at the top
- Hide event exclusive spawns
- Show event exclsuive spawns
--> List header still shows that list is sorted by Location descending, but event-exclusive spawn is at the bottom.

Finrod-Amandil commented on PokeOneWeb/Frontend/pokeoneweb/src/app/pages/pokemon-detail/pokemon-detail.component.html:
```diff
@@ -645,12 +645,26 @@ <h2 name="abilities">Abilities</h2>
+                </button>
+            </div>
+            <div *ngIf="model.areOnlyEventExclusiveSpawnsAvailable">
+                <p class="empty">This Pokémon is only obtainable while events are running in PokéOne, thus showing event-exclusive spawns.</p>
```
Change message to "This Pokémon was only available during events in the past". Change font color to red (color: var(--text-accent))

Finrod-Amandil commented on PokeOneWeb/Frontend/pokeoneweb/src/app/pages/pokemon-detail/pokemon-detail.component.ts:
```diff
@@ -46,14 +46,17 @@ export class PokemonDetailComponent implements OnInit {
+
+                    this.model.learnableMoves = this.model.pokemon.learnableMoves;
+                    
+                    if(this.model.areEventExclusiveSpawnsHidden) {
```
if-statement is redundant, as event-exclusive spawns are always hidden on page load.

Bischi1997 replied:
What if the default behavior would be changed in PokemonDetailModel?
Wouldn't you then have to change it manually in the init-function?

Finrod-Amandil commented on PokeOneWeb/Frontend/pokeoneweb/src/app/pages/pokemon-detail/pokemon-detail.component.ts:
```diff
@@ -167,7 +170,48 @@ export class PokemonDetailComponent implements OnInit {
+            let today = this.getTodaysDate().split("/");
+
+            for (let spawn of this.model.pokemon.spawns){
+                if(spawn.isEvent){
```
I would extract a method here, to reduce nesting. i.e. something like
```
if(isSpawnCurrentlyAvailable(spawn)) {
   this.model.visibleSpawns.push(spawn);
}
```

Finrod-Amandil commented on PokeOneWeb/Frontend/pokeoneweb/src/app/pages/pokemon-detail/pokemon-detail.component.ts:
```diff
@@ -167,7 +170,48 @@ export class PokemonDetailComponent implements OnInit {
+
+    public hideEventExclusiveSpawns() {
+        this.model.areEventExclusiveSpawnsHidden = true;
+        if(this.model.pokemon) {
```
reverse if-statement to reduce nesting
```
if (!this.model.pokemon) {
   return;
}
```

Finrod-Amandil commented on PokeOneWeb/Frontend/pokeoneweb/src/app/pages/pokemon-detail/pokemon-detail.component.ts:
```diff
@@ -167,7 +170,48 @@ export class PokemonDetailComponent implements OnInit {
+                    this.model.visibleSpawns.push(spawn);
+                }
+            }
+            if(this.model.visibleSpawns.length === 0){
```
Add a comment explaining this case.

Finrod-Amandil commented on PokeOneWeb/Frontend/pokeoneweb/src/app/pages/pokemon-detail/pokemon-detail.component.ts:
```diff
@@ -249,17 +293,35 @@ export class PokemonDetailComponent implements OnInit {
             });
     }
 
+    private convertDate(date: String) {
```
I would move these generic helper classes to a separate class / date service + dependency injection

Bischi1997 committed:
added requested changes

missing changes:
- https://github.com/Finrod-Amandil/PokeOneWeb/pull/129#discussion_r841230398
- https://github.com/Finrod-Amandil/PokeOneWeb/pull/129#discussion_r841231472

Bischi1997 committed:
implemented date service + dependency injection as requested

Bischi1997 committed:
added spy to jasmine

testing for newly created things still missing

Bischi1997 committed:
fixed codesmell and 2x bug found by sonarcloud

Finrod-Amandil commented on PokeOneWeb/Frontend/pokeoneweb/src/app/pages/pokemon-detail/pokemon-detail.component.ts:
```diff
@@ -167,7 +171,83 @@ export class PokemonDetailComponent implements OnInit {
+        this.checkAreNoEventSpawnsAvailable(this.model.pokemon.spawns);
+
+        for (let spawn of this.model.pokemon.spawns){
+            if(spawn.isEvent){
```
I would move the spawn.isEvent check to the extracted method, so that you only have the for loop and a single if-statement in here. Negate the check in the method, i.e.
```
if (!spawn.isEvent) {
  return true;
}

// Check event spawn
```

Finrod-Amandil commented on PokeOneWeb/Frontend/pokeoneweb/src/app/pages/pokemon-detail/pokemon-detail.component.ts:
```diff
@@ -46,14 +49,15 @@ export class PokemonDetailComponent implements OnInit {
-            });
+                    this.model.learnableMoves = this.model.pokemon.learnableMoves;
+                    
+                    this.hideEventExclusiveSpawns();
```
As answer to your question about the if-statement here: No, if the default behaviour changes, you would just have to call "showEventExclusiveSpawns()" here instead of "hide...". The initial value that is set in the model is irrelevant, as both methods (hide/show) update the boolean value anyway.

Bischi1997 committed:
updated according to feedback of reviewer

Bischi1997 committed:
added tests for date.service

Bischi1997 committed:
added some simple tests to spec-sheet

Bischi1997 committed:
added material styling for button

Bischi1997 committed:
added some tests for newly written functions

Bischi1997 committed:
added more tests
